File: packages/core-ts/src/oauth/provider.ts
/**
 * OAuth 2.1 Provider logic for TypeScript MCP servers.
 *
 * Implements the "MCP Server as Authorization Server" pattern.
 * The MCP Server issues JWTs, processes PKCE verification,
 * and uses the Relay Server purely as a UI consent transport.
 */

import { createHash, timingSafeEqual } from 'node:crypto'
import { createSession, pollForResult, type RelaySession } from '../relay/client.js'
import type { RelayConfigSchema } from '../schema/types.js'
import type { JWTIssuer } from './jwt-issuer.js'

export interface PreAuthSession {
  sessionId: string
  clientId: string
  redirectUri: string
  state: string
  codeChallenge: string
  codeChallengeMethod: string
  /** Serialized CryptoKeyPair for later credential retrieval */
  keyPairJwk: JsonWebKey
  passphrase: string
  expiresAt: number
}

export interface IOAuthSessionCache {
  save(session: PreAuthSession): void
  getAndDelete(sessionId: string): PreAuthSession | null
}

export class InMemoryAuthCache implements IOAuthSessionCache {
  private cache = new Map<string, PreAuthSession>()

  save(session: PreAuthSession): void {
    this.cache.set(session.sessionId, session)
    // Cleanup expired entries
    const now = Math.floor(Date.now() / 1000)
    for (const [id, sess] of this.cache) {
      if (sess.expiresAt < now) this.cache.delete(id)
    }
  }

  getAndDelete(sessionId: string): PreAuthSession | null {
    const sess = this.cache.get(sessionId)
    if (!sess) return null
    this.cache.delete(sessionId)
    if (sess.expiresAt < Math.floor(Date.now() / 1000)) return null
    return sess
  }
}

export interface OAuthProviderOptions {
  serverName: string
  relayBaseUrl: string
  relaySchema: RelayConfigSchema
  jwtIssuer: JWTIssuer
  cache?: IOAuthSessionCache
}

export class OAuthProvider {
  private serverName: string
  private relayBaseUrl: string
  private relaySchema: RelayConfigSchema
  private jwtIssuer: JWTIssuer
  private cache: IOAuthSessionCache

  constructor(options: OAuthProviderOptions) {
    this.serverName = options.serverName
    this.relayBaseUrl = options.relayBaseUrl
    this.relaySchema = options.relaySchema
    this.jwtIssuer = options.jwtIssuer
    this.cache = options.cache ?? new InMemoryAuthCache()
  }

  /**
   * Create a relay session and return the URL to redirect the user to.
   */
  async createAuthorizeRedirect(
    clientId: string,
    redirectUri: string,
    state: string,
    codeChallenge: string,
    codeChallengeMethod = 'S256'
  ): Promise<string> {
    const session = await createSession(this.relayBaseUrl, this.serverName, this.relaySchema)

    // Export private key for later credential retrieval
    const privateJwk = await crypto.subtle.exportKey('jwk', session.keyPair.privateKey)

    const preAuth: PreAuthSession = {
      sessionId: session.sessionId,
      clientId,
      redirectUri,
      state,
      codeChallenge,
      codeChallengeMethod,
      keyPairJwk: privateJwk,
      passphrase: session.passphrase,
      expiresAt: Math.floor(Date.now() / 1000) + 600 // 10 minutes
    }
    this.cache.save(preAuth)
    return session.relayUrl
  }

  /**
   * Exchange the authorization code (relay session ID) for an access_token.
   * PKCE verification is performed.
   *
   * @param code The authorization code from the client.
   * @param codeVerifier The PKCE code verifier.
   * @param userIdExtractor Function to derive a unique user_id from the credentials.
   * @returns Tuple of [access_token, credentials]
   */
  async exchangeCode(
    code: string,
    codeVerifier: string,
    userIdExtractor: (credentials: Record<string, string>) => string
  ): Promise<{ accessToken: string; credentials: Record<string, string> }> {
    const preAuth = this.cache.getAndDelete(code)
    if (!preAuth) throw new Error('invalid_grant: Expired or invalid code')

    // Verify PKCE
    if (preAuth.codeChallengeMethod === 'S256') {
      const digest = createHash('sha256').update(codeVerifier).digest('base64url')
      const expected = Buffer.from(digest)
      const actual = Buffer.from(preAuth.codeChallenge)
      if (expected.length !== actual.length || !timingSafeEqual(expected, actual)) {
        throw new Error('invalid_grant: PKCE verification failed')
      }
    } else if (preAuth.codeChallengeMethod === 'plain') {
      const expected = Buffer.from(codeVerifier)
      const actual = Buffer.from(preAuth.codeChallenge)
      if (expected.length !== actual.length || !timingSafeEqual(expected, actual)) {
        throw new Error('invalid_grant: PKCE plain verification failed')
      }
    } else {
      throw new Error('unsupported_challenge_method')
    }

    // Reconstruct key pair to decrypt relay credentials
    const privateKey = await crypto.subtle.importKey(
      'jwk',
      preAuth.keyPairJwk,
      { name: 'ECDH', namedCurve: 'P-256' },
      true,
      ['deriveKey', 'deriveBits']
    )
    const publicKey = await crypto.subtle.importKey(
      'jwk',
      { ...preAuth.keyPairJwk, d: undefined },
      { name: 'ECDH', namedCurve: 'P-256' },
      true,
      []
    )
    const reconstructedSession: RelaySession = {
      sessionId: preAuth.sessionId,
      keyPair: { privateKey, publicKey },
      passphrase: preAuth.passphrase,
      relayUrl: ''
    }

    // Poll for credentials (short timeout — form already submitted)
    const credentials = await pollForResult(this.relayBaseUrl, reconstructedSession, 1000, 10_000)

    // Extract unique user_id
    const userId = userIdExtractor(credentials)
    if (!userId) throw new Error('server_error: Unable to extract user_id from credentials')

    // Issue access token
    const accessToken = await this.jwtIssuer.issueAccessToken(userId)
    return { accessToken, credentials }
  }

  /**
   * Get RFC 8414 Authorization Server Metadata.
   */
  getMetadata(baseUrl: string): Record<string, unknown> {
    return {
      issuer: baseUrl,
      authorization_endpoint: `${baseUrl}/authorize`,
      token_endpoint: `${baseUrl}/token`,
      jwks_uri: `${baseUrl}/.well-known/jwks.json`,
      response_types_supported: ['code'],
      grant_types_supported: ['authorization_code'],
      code_challenge_methods_supported: ['S256', 'plain'],
      token_endpoint_auth_methods_supported: ['none']
    }
  }
}

File: packages/core-py/src/mcp_relay_core/oauth/provider.py
import base64
import hashlib
import hmac
import time
from collections.abc import Callable
from dataclasses import dataclass
from typing import Protocol

from mcp_relay_core.crypto.ecdh import export_private_key, import_private_key
from mcp_relay_core.relay.client import RelaySession, create_session, poll_for_result
from mcp_relay_core.schema.types import RelayConfigSchema

from .jwt_issuer import JWTIssuer

"""OAuth 2.1 Provider logic for MCP servers.

This implements the "MCP Server as Authorization Server" pattern.
The MCP Server issues the JWTs, processes the PKCE verification,
and uses the Relay Server purely as a UI consent transport.
"""


@dataclass
class PreAuthSession:
    """Pending auth before code exchange."""

    session_id: str
    client_id: str
    redirect_uri: str
    state: str
    code_challenge: str
    code_challenge_method: str
    private_key_b64: str
    passphrase: str
    expires_at: int


class IOAuthSessionCache(Protocol):
    """Cache for maintaining state between /authorize and /token endpoints."""

    def save(self, session: PreAuthSession) -> None: ...
    def get_and_delete(self, session_id: str) -> PreAuthSession | None: ...


class InMemoryAuthCache(IOAuthSessionCache):
    def __init__(self):
        self._cache: dict[str, PreAuthSession] = {}

    def save(self, session: PreAuthSession) -> None:
        self._cache[session.session_id] = session
        # Cleanup expired
        now = int(time.time())
        expired = [sid for sid, sess in self._cache.items() if sess.expires_at < now]
        for sid in expired:
            del self._cache[sid]

    def get_and_delete(self, session_id: str) -> PreAuthSession | None:
        if session_id in self._cache:
            sess = self._cache.pop(session_id)
            if sess.expires_at >= int(time.time()):
                return sess
        return None


class OAuthProvider:
    def __init__(
        self,
        server_name: str,
        relay_base_url: str,
        relay_schema: RelayConfigSchema,
        jwt_issuer: JWTIssuer,
        cache: IOAuthSessionCache | None = None,
    ):
        self.server_name = server_name
        self.relay_base_url = relay_base_url
        self.relay_schema = relay_schema
        self.jwt_issuer = jwt_issuer
        self.cache = cache or InMemoryAuthCache()

    async def create_authorize_redirect(
        self,
        client_id: str,
        redirect_uri: str,
        state: str,
        code_challenge: str,
        code_challenge_method: str = "S256",
    ) -> str:
        """Create a relay session and return the URL to redirect the user to."""

        # Call relay to create session, appending the oauthState param
        session = await create_session(
            self.relay_base_url,
            self.server_name,
            self.relay_schema,
        )

        # Store private key and passphrase temporarily
        pre_auth = PreAuthSession(
            session_id=session.session_id,
            client_id=client_id,
            redirect_uri=redirect_uri,
            state=state,
            code_challenge=code_challenge,
            code_challenge_method=code_challenge_method,
            private_key_b64=export_private_key(session.private_key),
            passphrase=session.passphrase,
            expires_at=int(time.time()) + 600,  # 10 mins
        )
        self.cache.save(pre_auth)
        return session.relay_url

    async def exchange_code(
        self,
        code: str,
        code_verifier: str,
        user_id_extractor: Callable[[dict], str],
    ) -> tuple[str, dict]:
        """
        Exchange the authorization code (which is the relay session_id)
        for an access_token. PKCE verification is performed.

        Args:
            code: The authorization code provided by the client.
            code_verifier: The PKCE code verifier.
            user_id_extractor: Function to derive a unique user_id from the credentials.

        Returns:
            Tuple of (access_token_jwt, credentials_dict)
        """
        pre_auth = self.cache.get_and_delete(code)
        if not pre_auth:
            raise ValueError("invalid_grant: Expired or invalid code")

        # Verify PKCE
        if pre_auth.code_challenge_method == "S256":
            digest = hashlib.sha256(code_verifier.encode("ascii")).digest()
            expected_challenge = (
                base64.urlsafe_b64encode(digest).decode("ascii").rstrip("=")
            )
            if not hmac.compare_digest(expected_challenge, pre_auth.code_challenge):
                raise ValueError("invalid_grant: PKCE verification failed")
        elif pre_auth.code_challenge_method == "plain":
            if not hmac.compare_digest(code_verifier, pre_auth.code_challenge):
                raise ValueError("invalid_grant: PKCE plain verification failed")
        else:
            raise ValueError("unsupported_challenge_method")

        # reconstruct RelaySession to decrypt
        relay_session = RelaySession(
            session_id=pre_auth.session_id,
            private_key=import_private_key(pre_auth.private_key_b64),
            public_key=None,  # Not needed for decryption
            passphrase=pre_auth.passphrase,
            relay_url="",
        )

        credentials = await poll_for_result(
            self.relay_base_url,
            relay_session,
            interval_s=1.0,
            timeout_s=10.0,
        )

        # Extract unique user_id
        user_id = user_id_extractor(credentials)
        if not user_id:
            raise ValueError("server_error: Unable to extract user_id from credentials")

        # Issue access token
        access_token = self.jwt_issuer.issue_access_token(sub=user_id)
        return access_token, credentials
